dithering_binning.py

```diff
@@ -0,0 +1,272 @@
+        """
+        if len(labels) == 0 or count == 0:
+            raise ValueError('Must have at least one Bin!')
+
+        for i in range(0, length):
+            self.label.append('')  # Fill in Label Empty so we can insert values in its according index
+
+        self.bin_count = count
+
+        for i in range(0, count):
+            empty_bin = Bucket(labels[i])
+            self.bins.append(empty_bin)
+
+    def _send_bucket(self, offset_value, split):
+        """Receive offset_value with split and calculate which bucket it belongs too"""
+
+        fit = 0
+        while offset_value >= split:
+            offset_value -= split
+            fit += 1
+        return fit
+
+    def distribution_by_value(self):
+        """Distribute coins into bins according to its value
+
+        We use the range of min and max to create a split number by value per bucket. The fit is made by calling
+        _send_bin, which will keep reducing the fit number by split to find the appropriate bucket it belongs into
+
+        Since every index is unique, we will use that as key for coins to store it in bins.
+        """
+
+        split = int((1 + self.max_value - self.min_value) / self.bin_count)
+
+        for index, coin in enumerate(self.coin_list):
+            if not isinstance(coin.value, int):
+                self.label[index] = 'NaN'
+                continue
+
+            offset_value = int(coin.value - self.min_value)
+            fit = self._send_bucket(offset_value, split)
+
+            if fit >= self.bin_count:  # Edge Case: If fit hits the max value, put in last bucket
+                self.bins[self.bin_count - 1].add_coin(coin, index)
+            else:
+                self.bins[fit].add_coin(coin, index)
+
+    def dithering_balance(self):
+        """Balance the weight in each bucket by moving coins around after coins are distributed in bins already.
+
+        Using Dithering, we distribute min/max value of each bins and move around coins with random weight chosen until
+        the weight in the bucket are close to being evenly distributed. -- Start with in-order then reverse for
+        full balance. Each run, we keep pushing to the next bucket until its weight is less than the threshold.
+
+        threshold
+            1. Weight / number of bins: the least amount of weight in each bucket
+            2. Average weight in each coin: To not over add coins in a bucket accounting with average value.
+            3. Threshold >= : threshold is naturally bigger than a split even weight for each bucket, so we use = sign
+            to favor more balance/dithering, as it would be a ~5050 anyways for next value to make it go below split
+            weight
+        """
+
+        split_weight = int(self.total_weight / self.bin_count)  # All bins weight is evenly split
+        average_weight = int(self.total_weight / len(self.coin_list))  # Bins average weight per coin
```

Comment:
k you got a lot of these unnecessary ints lol.
You only need to cast int prior to a mathematic operation, not after